### DISCUSSION TITLE
Fast Pattern Matching across 4 or 5 different time series

### DISCUSSION THREAD

**Discussion by Dwarak-PCI** [TDAmeritrade/stumpy#541] [Q&A] [answered]
Given a time series with historical temperature data - obs(t) and another time series with temperature forecast - yhat(t), I can follow this tutorial (https://github.com/TDAmeritrade/stumpy/blob/main/docs/Tutorial_Pattern_Matching.ipynb) to identify any patterns observed in the historical temperature data that matches the temperature forecast time series. This is very impressive.

I have another use case. Lets say, I have 4 historical time series data namely temp_obs(t), humidity_obs(t), windspeed_obs(t), precipitation _obs(t) and 4 forecast time series namely temp_forecast(t), humidity_forecast(t), windspeed_forecast(t), precipitation _forecast(t). I would like to identify dates/days where forecast time series pattern matches historical time series. Is there any existing tutorial that can help me with this by finding patterns across 4 different time series and providing a single matrix profile?

**Comment by JaKasb** [comment, marked as answer]
If you want to search for motifs in the same physical quantity, then you want to perform an AB-Join.

`stumpy.stump(T_A = temp_obs, m=???, T_B=temp_forecast, ignore_trivial=False, normalize=???)`
The parameters (m, normalize) are domain dependent.

Maybe the `ostinato` algorithm could be useful (see https://stumpy.readthedocs.io/en/latest/Tutorial_Consensus_Motif.html).

Another data mining task is to search for [multidimensional motifs](https://stumpy.readthedocs.io/en/latest/Tutorial_Multidimensional_Motif_Discovery.html).
If the time-axis is synchronized across all t_obs, you can interpret the recordings as a single 4-D timeseries.
If you interpret the problem as 4D-Timeseries, you also need to analyze the interpretation of a 4D-Distance-Measure for pattern-similarity.
The problem for 4D physical quantities is, that it may be nonsensical to individually z-normalize each dimension.
Z-Normalization destroys any knowledge about the physical magnitude.

**Reply by seanlaw** [reply]
Assuming that your data are all time aligned (i.e.,  the length  are identical), you may also consider concatenating your observed data to your forecast data:

```
import numpy as np
import stumpy

temp_obs = np.random.rand(100)
temp_forecast = np.random.rand(1000)
humid_obs = np.random.rand(100)
humid_forecast = np.random.rand(1000)

obs = np.vstack([temp_obs, temp_obs])
forecast = np.vstack([temp_forecast, temp_forecast])
nan = np.full((2, 1), np.nan)
T = np.hstack([obs, nan, forecast])  # Add nan between observed and forecast to ensure that no motifs are found straddling
```

And then you may be  able use `mstump` to compute the multi-dimensional matrix profile
```
mp, indices = mstump(T, m)
```

However, as @JaKasb pointed out, it is up to you to determine how to interpret these results as the discovered motifs may be meaningless.

**Comment by Dwarak-PCI** [comment]
Thank you very much for the explanation. @seanlaw @JaKasb
